arrow-docs/docs/io/README.md

```diff
@@ -12,3 +12,206 @@ Arrow has adopted suspend as system to model monadic computations and offers the
+
+We can use extension functions to do functional dependency injection with similar semantics as `Reader` or `Kleisli`.
+They allow us to elegantly define `syntax` for a certain type. Let's see a simple example.
+
+Let's reuse our previous domain of`User`, `ProcessedUser`, but let's introduce `Repo` and `Persistence` layers to mimick what could be a small app with a couple layers.
+
+```kotlin:ank
+interface Repo {
+    suspend fun fetchUsers(): List<User>
+}
+
+interface Persistence {
+    suspend fun User.process(): Either<ProcessingError, ProcessedUser>
+
+    suspend fun List<User>.process(): Either<ProcessingError, List<ProcessedUser>> =
+        either { map { !it.process() } }
+}
+```
+
+Given the above defined layers we can easily compose them by creating a product which implements the dependencies by delegation.
+This can easily be done manually or with the help of a framework like Dagger to do this wiring automatically.
+
+```kotlin:ank
+class DataModule(
+    persistence: Persistence,
+    repo: Repo
+) : Persistence by persistence, Repo by repo
+```
+
+We can also define top-level functions based on constraints on the receiver.
+Here we define `getProcessUsers` which can only be called where `R` is both `Repo` and `Persistence`.
+
+```kotlin:ank
+/**
+ * Generic top-level function based on syntax enabled by [Persistence] & [Repo] constraint
+ */
+suspend fun <R> R.getProcessUsers(): Either<ProcessingError, List<ProcessedUser>>
+        where R : Repo,
+              R : Persistence = fetchUsers().process()
+```
+
+### Performance
+
+`suspend` is extremely fast in comparison to `IO<A>`, since `IO<A>` is build at runtime and `suspend` is build by the compiler.
+
+Let's take our previous example from ergonomics:
+ 
+```kotlin:ank
+import arrow.fx.IO
+
+fun number(): IO<Int> = IO.just(1)
+
+fun triple(): IO<Triple<Int, Int, Int>> =
+  number().flatMap { a ->
+    number().flatMap { b ->
+      number().map { c ->
+        Triple(a, b, c)
+      }
+    }
+  }
+```
+
+If we translate this piece of code to the `data class` it uses it results in the following:
+
+```kotlin
+fun number(): IO<Int> = IO.Just(1)
+
+fun triple(): IO<Triple<Int, Int, Int>> =
+  IO.Bind(IO.Just(1)) { a ->
+    IO.Bind(IO.Just(1)) { b ->
+      IO.Map(IO.Just(1)) { c -> Triple(a, b, c) }
+    }
+  }
+```
+
+So we can see that for this simple program we allocated 6 `IO` cases of the `IO` `sealed class`.
+This is necessary so when you `unsafeRun` the `IO` program, it can be inspected to find out which branch of code needs to run. 
+
+In contrast to this `suspend` can simply be wired by the Kotlin compiler, which eliminates the need for this `sealed class` and allocations.
+The Kotlin compiler rewrites the suspend program to a super fast runtime which uses a switch table and mutable state machine to run the `suspend` program.
+Furthermore, the Kotlin compiler applies a bunch of optimisations that improve the speed of `suspend` even further and certains JVM .
```

Comment:
```suggestion
Furthermore, the Kotlin compiler applies other optimisations focused on the speed of `suspend` and memory usage of suspension making it suitable for hotspots and places where otherwise allocations or ADT based data types are not an option.
```